packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+                attr.name
+              ];
+              const attrNameStartOffset = attrSourceLocation.startOffset;
+              const attrEndOffset = attrSourceLocation.endOffset;
+              flushTo(attrNameStartOffset);
+              ops.push({
+                type: 'attribute-part',
+                index: nodeIndex,
+                name,
+                strings,
+                tagName,
+                useCustomElementInstance: ctor !== undefined,
+              });
+              skipTo(attrEndOffset);
+            } else if (node.isDefinedCustomElement) {
+              // We will wait until after connectedCallback() and render all
+              // custom element attributes then
+              const attrSourceLocation = node.sourceCodeLocation!.attrs[
+                attr.name
+              ];
+              flushTo(attrSourceLocation.startOffset);
+              skipTo(attrSourceLocation.endOffset);
+            }
+          }
+        }
+
+        if (writeTag) {
+          if (node.isDefinedCustomElement) {
+            flushTo(node.sourceCodeLocation!.startTag.endOffset - 1);
+            ops.push({
+              type: 'custom-element-attributes',
+            });
+            flush('>');
+            skipTo(node.sourceCodeLocation!.startTag.endOffset);
+          } else {
+            flushTo(node.sourceCodeLocation!.startTag.endOffset);
+          }
+        }
+
+        if (boundAttrsCount > 0) {
+          flush(`<!--lit-bindings ${nodeIndex}-->`);
+        }
+
+        if (ctor !== undefined) {
+          ops.push({
+            type: 'custom-element-children',
+          });
+        }
+      }
+      nodeIndex++;
+    },
+    post(node) {
+      if (isElement(node) && node.isDefinedCustomElement) {
+        ops.push({
+          type: 'custom-element-close',
+        });
+      }
+    },
+  });
+  flushTo();
+  const t = {ops, attrNames};
+  templateCache.set(result.strings, t);
+  return t;
+};
+
+export type RenderInfo = {
+  customElementInstanceStack: Array<ElementRenderer | undefined>;
+};
+
+declare global {
+  interface Array<T> {
+    flat(depth: number): Array<T>;
+  }
+}
+
+export function* render(value: unknown): IterableIterator<string> {
+  yield* renderValue(value, {customElementInstanceStack: []});
+}
+
+export function* renderValue(
+  value: unknown,
+  renderInfo: RenderInfo
+): IterableIterator<string> {
+  if (isRenderLightDirective(value)) {
+    // If a value was produced with renderLight(), we want to call and render
+    // the renderLight() method.
+    const instance = getLast(renderInfo.customElementInstanceStack);
+    if (instance !== undefined) {
+      yield* instance.renderLight(renderInfo);
+    }
+    value = null;
+  } else if (value != null && (value as DirectiveResult)._$litDirective$) {
+    const directive = (value as DirectiveResult)._$litDirective$;
+    value = new directive({type: NODE_PART}).render(
```

Comment:
Add comment that we're calling `render` and not `update` as the client does by choice because `render` is expected to have a renderable output.

Reply:
Done.